File: jetweb-core/core/js/src/pikr.js
export const PSILen = 32;
export const SFILen = 16;
export const TSILen = 16;
export const QIILen = 32;
export const QRILen = 8;
export const HashLen = 32;

export function newIdentity5(psi, sfi, tsi, qii, qri) {
  if (
    psi.length !== PSILen ||
    sfi.length !== SFILen ||
    tsi.length !== TSILen ||
    qii.length !== QIILen ||
    qri.length !== QRILen
  ) {
    throw new Error("Invalid dimension length for Identity5");
  }
  return { psi, sfi, tsi, qii, qri };
}

import { blake2b } from "blakejs";

async function hashBlake256(data) {
  return blake2b(data, null, HashLen);
}

export async function recoveryMatrix(id) {
  return {
    k1: await hashBlake256(id.psi),
    k2: await hashBlake256(id.sfi),
    k3: await hashBlake256(id.tsi),
    k4: await hashBlake256(id.qii),
    k5: await hashBlake256(id.qri),
  };
}

export async function sovereignKey(rm) {
  const buf = new Uint8Array(HashLen * 5);
  buf.set(rm.k1, 0);
  buf.set(rm.k2, HashLen);
  buf.set(rm.k3, HashLen * 2);
  buf.set(rm.k4, HashLen * 3);
  buf.set(rm.k5, HashLen * 4);
  return hashBlake256(buf);
}

function toHex(bytes) {
  return Array.from(bytes)
    .map(b => b.toString(16).padStart(2, "0"))
    .join("");
}

function shortHex(bytes, n) {
  const h = toHex(bytes);
  return h.slice(0, n);
}

export function identityName(id) {
  return `PI-${shortHex(id.psi, 6)}-${shortHex(id.qri, 2)}`;
}

export function lineageName(id, generation = 0) {
  const gen = Math.max(0, generation).toString().padStart(3, "0");
  return `LN-${shortHex(id.psi, 6)}-${shortHex(id.qii, 4)}-${gen}`;
}

export function recoveryName(dim, k) {
  return `REC-${dim}-${shortHex(k, 4)}`;
}

File: jetweb-core/core/go/pikr/pikr.go
package pikr

import (
	"encoding/hex"
	"errors"
	"fmt"

	"golang.org/x/crypto/blake2b"
)

const (
	PSILen  = 32
	SFILen  = 16
	TSILen  = 16
	QIILen  = 32
	QRILen  = 8
	ID5Len  = PSILen + SFILen + TSILen + QIILen + QRILen
	HashLen = 32
)

var ErrInvalidLength = errors.New("invalid dimension length for Identity5")

type Identity5 struct {
	PSI []byte // 32
	SFI []byte // 16
	TSI []byte // 16
	QII []byte // 32
	QRI []byte // 8
}

type RecoveryMatrix struct {
	K1 []byte
	K2 []byte
	K3 []byte
	K4 []byte
	K5 []byte
}

func NewIdentity5(psi, sfi, tsi, qii, qri []byte) (*Identity5, error) {
	if len(psi) != PSILen || len(sfi) != SFILen || len(tsi) != TSILen ||
		len(qii) != QIILen || len(qri) != QRILen {
		return nil, ErrInvalidLength
	}
	return &Identity5{psi, sfi, tsi, qii, qri}, nil
}

func (id *Identity5) ID5() []byte {
	out := make([]byte, 0, ID5Len)
	out = append(out, id.PSI...)
	out = append(out, id.SFI...)
	out = append(out, id.TSI...)
	out = append(out, id.QII...)
	out = append(out, id.QRI...)
	return out
}

func hashBlake256(data []byte) []byte {
	h, _ := blake2b.New256(nil)
	h.Write(data)
	return h.Sum(nil)
}

func (id *Identity5) RecoveryMatrix() *RecoveryMatrix {
	return &RecoveryMatrix{
		K1: hashBlake256(id.PSI),
		K2: hashBlake256(id.SFI),
		K3: hashBlake256(id.TSI),
		K4: hashBlake256(id.QII),
		K5: hashBlake256(id.QRI),
	}
}

func (rm *RecoveryMatrix) SovereignKey() []byte {
	buf := make([]byte, 0, HashLen*5)
	buf = append(buf, rm.K1...)
	buf = append(buf, rm.K2...)
	buf = append(buf, rm.K3...)
	buf = append(buf, rm.K4...)
	buf = append(buf, rm.K5...)
	return hashBlake256(buf)
}

// Nomenclature

func shortHex(b []byte, n int) string {
	h := hex.EncodeToString(b)
	if len(h) < n {
		return h
	}
	return h[:n]
}

func (id *Identity5) IdentityName() string {
	return "PI-" + shortHex(id.PSI, 6) + "-" + shortHex(id.QRI, 2)
}

func (id *Identity5) LineageName(generation int) string {
	return "LN-" + shortHex(id.PSI, 6) + "-" + shortHex(id.QII, 4) +
		"-" + fmt.Sprintf("%03d", max(0, generation))
}

func (rm *RecoveryMatrix) RecoveryName(dim string, k []byte) string {
	return "REC-" + dim + "-" + shortHex(k, 4)
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
